components/brave_rewards/core/common/legacy_callback_helpers.h

```diff
@@ -0,0 +1,32 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_REWARDS_CORE_COMMON_LEGACY_CALLBACK_HELPERS_H_
+#define BRAVE_COMPONENTS_BRAVE_REWARDS_CORE_COMMON_LEGACY_CALLBACK_HELPERS_H_
+
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "base/functional/callback.h"
+
+namespace ledger {
+
+// Converts a `OnceCallback` into an equivalent `std::function`. This adapter
+// should only be used to interface with legacy code that uses `std::function`
+// for callbacks. Use `OnceCallback` for all new code.
+template <typename... Args>
+std::function<void(Args...)> ToLegacyCallback(
+    base::OnceCallback<void(Args...)> callback) {
+  auto shared_callback =
```

Comment:
It would be nice if `shared_callback` was created in the capture list to avoid having to do an atomic increment of the reference counter in `std::shared_ptr<>` when capturing `shared_callback` by value (`std::move<>()`ing it is an alternative, but it incurs the cost of an unnecessary move-constructor call).

Also, this doesn't do what you'd expect. `args` are not universal references here, as there's no type deduction taking place - `Args&&...` are already determined by `base::OnceCalback<void(Args...)>`.

Could be something like:
```
template <typename... Args>
auto ToLegacyCallback(base::OnceCallback<void(Args...)> callback) {
  return [shared_callback = std::make_shared<decltype(callback)>(
              std::move(callback))](auto&&... args) {
    std::move(*shared_callback).Run(std::forward<decltype(args)>(args)...);
  };
}
```

Reply:
It's unfortunate that long capture lists (and lambdas in general) are formatted so poorly by clang-format. I dislike having to choose between minor technical issues and readability.

Regarding `auto&&`, I though of that, but my thinking was that since we are providing a fully-specified  `OnceCallback`, the resulting `std::function` should accept those arguments and nothing else. No?

Reply:
I'll have to think that through, but something like this should work:
```
template <typename... Args>
auto ToLegacyCallback(base::OnceCallback<void(Args...)> callback) {
  return [shared_callback = std::make_shared<decltype(callback)>(
              std::move(callback))](Args... args) {
    std::move(*shared_callback).Run(std::forward<Args>(args)...);
  };
}
```
despite the unorthodox `std::forward<>()` usage.

Reply:
- `std::string`         `==>` gets cast to rvalue, ends up being moved        `==>` OK
- `std::string const`   `==>` gets cast to `const` rvalue, ends up being copied `==>` OK
- `std::string&`        `==>` gets cast to lvalue, ends up being copied       `==>` OK
- `std::string const&`  `==>` gets cast to `const` lvalue, ends up being copied `==>` OK
- `std::string&&`       `==>` gets cast to rvalue, ends up being moved        `==>` OK
- `std::string const&&` `==>` gets cast to `const` rvalue, ends up being copied `==>` OK

Seems right.

Reply:
Updated to use both your original suggestions. Thanks!